teps/0023-implicit-mapping.md

```diff
@@ -0,0 +1,358 @@
+passed through the PipelineRun -> Pipeline -> Tasks (often completely
+unmodified), and adding additional params does not inherently modify Task
+behavior if they happen to be unused.
+
+### Goal
+
+- Simplify authored Pipeline YAML by allowing users to omit parameters from
+  embedded resources.
+- Keep Pipeline API the same w.r.t. the Pipeline controller.
+
+### Non Goals
+
+- Support implicit mapping for PipelineResources, Workspaces, Task/PipelineRefs.
+
+## Requirements
+
+<!--
+List the requirements for this TEP.
+-->
+
+1. Resource authors can optionally omit parameters that can be inferred from a
+   parent scope.
+2. Resources should be fully resolved before they are stored in etcd.
+3. The Pipeline controller should only ever work with fully resolved resources.
+
+## Proposal
+
+We propose allowing for implicit parameters in a resource spec, by passing
+through parameters from parents. For example:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    params:
+      - name: MESSAGE
+        type: string
+    tasks:
+      - name: echo-message
+        taskSpec:
+          params:
+            - name: MESSAGE
+              type: string
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+        params:
+          - name: MESSAGE
+            value: $(params.MESSAGE)
+  params:
+    - name: MESSAGE
+      value: "Good Morning!"
+```
+
+We want users to be able to optionally shorten this config by removing param
+definitions in embedded specs:
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: PipelineRun
+metadata:
+  name: pipelinerun-with-taskspec-to-echo-message
+spec:
+  pipelineSpec:
+    tasks:
+      - name: echo-message
+        taskSpec:
+          steps:
+            - name: echo
+              image: ubuntu
+              script: |
+                #!/usr/bin/env bash
+                echo "$(params.MESSAGE)"
+  params:
+    - name: MESSAGE
+      value: "Good Morning!"
+```
+
+With the shortened syntax, we can derive the same information we need from the
+top-most parameter definition -
+
+1. We need a param named `MESSAGE`
+2. The param value is of type `string`
+3. No default value is required since we already have a value.
+
+Since these are all embedded specs within the same resource, our expectation is
+that this shortened syntax will not take away from the context knowledge of
+parameters and how they should be used. A good analogy here is variable scoping
+found in most programming languages - e.g.
+
+```go
+x := "Hi!"
+func() {
+  // x is still reachable, even though it's not explicitly defined parameter.
+  fmt.Println(x)
+}()
+```
+
+### Admission Controller
+
+This logic should live within the Pipelines mutating admission controller.
```

Comment:
with the OCI bundles the tekton controller isn't aware of them until runtime, i.e. when the PipelineRun/TaskRun is executing. so any validation that is happening when the bundle is created would either happen then OR the tool that is used to create the bundle

looking at the CLI TEP (https://github.com/tektoncd/community/blob/main/teps/0031-tekton-bundles-cli.md#requirements) it has a requirement "Only valid Tekton Bundles are produced." so I'm guessing the CLI does validation of the specs

Reply:
What do you think about adding a requirement that this approach is only undertaken for runtime types like PipelineRun and TaskRun? Maybe it doesn't make sense here since this proposal isn't about either of those 🤔  but i think it's worth recording in here somewhere (maybe in risks and mitigations? or some kind of future considerations section?)

Reply:
In my mind, PipelineRuns and Pipelines should have consistent behavior, since they are both using the same underlying PipelineSpec type. It feels strange to me if we had different behavior based on how you specified the same config.

Digging into the code, I think what makes bundles tricky is that it does not use the mutation/validation code, it just [decodes the objects and adds them as layers](https://github.com/tektoncd/cli/blob/892cc2caa5a330a007814f28feaad94046ef1d1e/pkg/bundle/builder.go#L31). So it's valid in the sense that it's structurally correct, but there might be other logical differences that make it different. If we add calls to `SetDefaults` and `Validate` on each decode, I think it should give us similar behavior. /cc @pierretasci in case you have any thoughts here.

I'll add an item for the next API WG for more discussion.

Reply:
Definitely worth discussing at API WG. From my perspective, the point of any remote resolution is just to be a different place to fetch a Pipeline, Task, etc. and any and all validation Tekton would normally do should continue to be done. Billy is correct that we are only validating the Bundle structure today but as we add more remote locations to fetch resources, validating their structure is entirely valid IMO.